### PR TITLE
kernel: Allow late processing of timeouts

### LINKED ISSUE #5401: delta_ticks_from_prev become negative and waiting tasks never scheduled
After a suspend period, if you wake up late and announce to the kernel a greater number of ticks  than the requested suspend period, timeouts are never processed in _handling_timeouts function because delta_ticks_from_prev become negative. 

To resolved this issue, i have make the following modification with sucess. Is anyone seeing a problem with this change?

```diff
kernel/sys_clock.c
@@ -240,7 +240,10 @@ static inline void handle_timeouts(s32_t ticks)
return;
}
- head->delta_ticks_from_prev -= ticks;
+ if (head->delta_ticks_from_prev <= ticks)
+ head->delta_ticks_from_prev = 0;
+ else
+ head->delta_ticks_from_prev -= ticks;
/*
* Dequeue all expired timeouts from _timeout_q, relieving irq lock

```

### PR DESCRIPTION
**Pull request by ghost**
This change proposes to handle the case where the handle_timeouts
function is called after a number of ticks greater than the first
timeout delta of the _timeout_q list. In the current implementation if
the case occurs, after subtracting the number of ticks the
delta_ticks_from_prev field becomes negative and the first timeout is
never processed. It is therefore necessary to treat this case and to
prevent delta_ticks_from_prev from becoming negative. Moreover, the lag
produced by the initial delay must also be applied to following timeouts
by browsing the list until it was entirely consumed.

Fixes #5401

### CODE REVIEW

**Comment by pfalcon** [comment]
@greenhandholman : Thanks for you patch! I'm sure, you would get more/earlier responses if the PR description was not empty. Please consider editing it and adding the relevant content. Thanks.

**Comment by pfalcon** [comment]
Also, I'd personally suggest to accompanies changes to the kernel with a testcase. It should fail without your patch, and not fail with it, clearly showing that there was an issue before the patch, and avoiding regressions to the introduced functionality in the future. Thanks.

**Review by andyross — COMMENTED** [review]
So... why isn't this just a signedness bug that can be treated with a lot less code?

I mean, I buy the description (i.e. when the tick count overshoots the first timeout, we do the wrong thing), I just don't understand why the fix needs to add so much code, and suspect that a much simpler trick might be available.

**Comment by greenhandholman** [comment]
Effectively this could be much simpler if the deadline for each time out in the queue were absolute, but in zephyr time out queue implementation each deadlines are relative to the previous timeout. Due to this we needs to browse the list to correct deadline of time out. That make code more complex but this is necessary.

File: kernel/sys_clock.c [hunk 1]
```diff
@@ -233,6 +256,7 @@ static inline void handle_timeouts(s32_t ticks)
 	_handle_expired_timeouts(&expired);
 
 	_handling_timeouts = 0;
+
```
Comment (andrewboie) [inline kernel/sys_clock.c]: stray whitespace

**Review by tyrantlizard77 — APPROVED** [review]
This LGTM.
But let's make sure you and @andyross are on the same page before merging.

File: kernel/sys_clock.c (hunk 1 above)
Comment (greenhandholman) [reply kernel/sys_clock.c]: Stray whitespace removed.
Thanks !

**Comment by andyross** [comment]
I'm fine merging this.  We can always hunt for cleanup opportunities later.

**Pull request merged**